flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -2508,4 +2538,92 @@ public void awaitRunning() throws InterruptedException {
             runningLatch.await();
         }
     }
+
+    /**
+     * A {@link StreamTask} that register a single timer that waits for a cancellation and then
+     * emits some data. The assumption is that output remains available until {@link
+     * AbstractInvokable#getTerminationFuturex()} is completed (even after cancellation). Public
+     * access to allow reflection in {@link Task}.
+     */
+    public static class StreamTaskWithBlockingTimer extends StreamTask {
+        static volatile CompletableFuture<Void> timerStarted;
+        static volatile CompletableFuture<Void> timerFinished;
+        static volatile CompletableFuture<Void> invokableCancelled;
+
+        public static void reset() {
+            timerStarted = new CompletableFuture<>();
+            timerFinished = new CompletableFuture<>();
+            invokableCancelled = new CompletableFuture<>();
+        }
+
+        // public access to allow reflection in Task
+        public StreamTaskWithBlockingTimer(Environment env) throws Exception {
+            super(env);
+            super.inputProcessor = getInputProcessor();
+            getCancelables().registerCloseable(() -> invokableCancelled.complete(null));
```

Comment:
How about we put the `invokableCancelled.complete(null)` into `cancelTask`? My concern here is that we apply an additional contract that the cancellables are closed in the `cancel` method.  (Which honestly I am not sure if it is correct as we do it a second time in the `cleanUpInvoke`, a different story though).

Reply:
I share your concern about the implicit contract. On the other hand, completing the future immediately will allow the timer to emit earlier and therefore reduce test sensitivity. I'm fine with both options, so I'll put it into `cancelTask` as you suggested.